rolldice.6
.\" Man file for rolldice(6) - v1.11 - 01 Feb 2010
.\" (c) Stevie Strickland, 1999-2012
.\"
.TH ROLLDICE 6 "01 Feb 2010" Linux
.SH NAME
rolldice \- rolls virtual dice
.SH SYNOPSIS
.B rolldice [
.I options
.B ]
.I [dice_string
.I [dice_string
.I ...]]
.SH DESCRIPTION
.B rolldice
rolls virtual dice.  The dice strings passed on the command line contain
information on the dice to roll in a format comparable to the format used in
fantasy role playing games.
.P
If no dice strings are provided as command line arguments,
.B rolldice
uses stdin as input.
.SH OPTIONS
.RS
.IP -h,--help
returns the usage of diceroll
.IP -v,--version
returns the version of diceroll 
.IP -r,--random
uses /dev/random for random number generating
.IP -u,--urandom
uses /dev/urandom for random number generating (default)
.IP -s,--separate
prints out the result of each individual die separately, as well as the
operations and totals 
.RE
.SH "DICE STRING FORMAT" 
The dice string uses the following format:
.RS
.IP {#x}{#}d[#|%]{*#}{+/-#}{s#}
.RE
.P
The dice string doesn't have to be in the format that I outlined above,
but this is the nicest order for me.  It will try to parse any different
string containing the same sections in the best way it can, and will
throw out anything that isn't one of the sections below.
.P Now, to break this format down section by section:
.RS
.IP {#}d[#|%]
This is the only required part of the string, and now really isn't even
required, as it defaults to 1d6 if not found.  The first number is the
number of dice to roll, and the second number is the number of sides the
dice have.  The numbers rolled on each die are then added up and given
as the result.  Hence 3d6 means "roll three six-sided dice, add them 
together and return the result".  If the first number is left out, then
the number of dice defaults to 1.  If the second number is not a number,
but a percentage sign, then the number of sides becomes 100 (for a
percentage roll).
.IP {#x}
This number tells how many times to roll.  For example, if you want to roll 3
6-sided dice 6 times, you would say 6x3d6.  This would return six numbers,
corresponding to the six different rolls.
.IP {*#}
This number tells how many times to multiply the result of each roll.
3d6*100 would return a number in the range of 300-1800, since 3-18 is the
range resulting from 3d6, which is then multipled by 100.
.IP {+/-#}
This number is the modifier to be added or subtracted, depending on the sign,
from each roll.  1d4+1 would result in a range from 2-5 (1-4 for the die, plus
1).  This step is handled *after* the multiplication modifier.
.IP {s#}
This number tells the program to drop the # lowest dice rolls.  This step is
handled *before* the multiplication modifier.
.RE
.P Any combination of the optional parts of the string may be used, but only
in the order show above.  For an extreme example, "3x4d6*5+1s2" would roll 
four six-sided dice, drop the lowest two, multiply the result by 5, add 1 to 
that, and repeat the process two more times, for a total of three results.
.SH DIAGNOSTICS
The following error messages may appear on STDERR:

Failure in getting dice attributes
.RS 
Either memory could not be allocated while parsing the string passed to the
.B rolldice
program or an invalid dice string was entered.
.RE
Failure in rolling dice
.RS 
Memory could not be allocated while actually rolling the dice.
.RE
Unknown option
.RS
Take a wild guess ;)
.RE
.SH BUGS
I haven't noticed any yet... 
.B rolldice
is a quick hack to provide a virtual dice roller for FRPGs that I play.
Any improvements are greatly welcome, as I am not the most experienced C
coder...
.SH AUTHOR
Stevie Strickland <sstrickl@ccs.neu.edu>
.SH VERSION
1.11 - 01 Feb 2010


Changelog
v1.11 - 01 Feb 2010
  - Only updating URLs/email addresses.  No bugfixes in this revision,
    though I'll get around to soon.

v1.10 - 25 Nov 2001
  - Forgot to add a couple of entries to the changelogs, I see.  Added them
    back in.
  - Truthfully, eventually I should rewrite this code completely, as I did
    originally write it before I really knew how one should write code like
    this.  Perhaps over the winter break.
  - This version fixes the problem of urandom always being opened, thanks
    to initializing the (u)random file before reading in the command line
    options.  Fixed that, plus replaced my option parsing code with a use
    of getopt_long().

v1.9 - 10 Mar 2001
  - Never checked result of parse_string for unsuccessful parses.  Fixed.

v1.8 - 17 Oct 1999
  - Segfaulted when one didn't specify how many dice to roll (should default
    to one).  Fixed.

v1.7 - 28 Mar 1999
  - Changed dice string parser code, now more to my liking :)
  - Added ability to use multiple dice strings

v1.6 - 27 Mar 1999
  - Added ability to drop a number of lowest die rolls and made reading
    from /dev/urandom the default.  In fact, there is no longer a method
    that uses the rand() function.  If adding this feature back is needed,
    then just contact me and I'll work it back in.

v1.5 - 10 Mar 1999
  - Added ability to print out each individual die roll and added the long
    version of each of the options

v1.4 - 12 Feb 1999
  - Added ability to use "d#" type constructions and "d%" for percentages

v1.3 - 27 Jan 1999
  - Fixed documentation, both in getting rid of childish flames and in
    correcting the dates in the changelog (and anywhere else they might
    be wrong :)

v1.2 - 27 Jan 1999
  - added /dev/(u)random code

v1.1 - 20 Jan 1999
  - Fixed bug involving overflowing and underflowingi because of
    too large or negative dice attributes

v1.0 - 19 Jan 1999
  - Released rolldice


rolldice.c
/**
 * rolldice.c - v1.11 - 01 Feb 2010
 * (c) Stevie Strickland, 1999-2012
 * 
 * This program has been placed under the GPL.  Any bugfixes or enhancements
 * will be greatly appreciated :)
 *
 * Stevie Strickland - sstrickl@ccs.neu.edu
 */

#include "rolldice.h"

// File pointer for random device
static FILE* ran_dev;

void init_random(rand_type rand_file) {
    if(rand_file == RANDOM) {
	if((ran_dev = fopen("/dev/random", "r")) == NULL) {
	    fprintf(stderr, "Error in opening /dev/random!\n");
	    exit(EXIT_FAILURE);
	}
    }
    else if((ran_dev = fopen("/dev/urandom", "r")) == NULL) {
	fprintf(stderr, "Error in opening /dev/urandom!\n");
	exit(EXIT_FAILURE);
    }
}

static int get_random(int sides) {
    unsigned int ret_value;
  
    if(!(fread(&ret_value, sizeof(unsigned int), 1, ran_dev) == 1)) {
	fprintf(stderr, "Error in reading random device!\n");
	exit(EXIT_FAILURE);
    }
    return (int)(ret_value % sides);
}

/* rolldie() - Rolls a single die
 *
 * Parameters: int num_sides - number of sides of the die to roll
 * Returns: int - the result of the roll
 */
int rolldie ( int num_sides ) {

    return (1 + get_random(num_sides));
}

/* parse_string() - Parses a string for dice rolling attributes
 *
 * Parameters: char *dice_string - string to parse
 * Returns: int * - array of nums describing the different aspects of the 
 * dice to be rolled
 */
int *parse_string(char *dice_string) {
    int temp_int = -1, *dice_nums;

    if((dice_nums = malloc ( DICE_ARRAY_SIZE * sizeof(int))) == NULL){
        perror("rolldice");
    	return NULL;
    }

    dice_nums[NUM_ROLLS] = 1;
    dice_nums[NUM_DICE] = 1;
    dice_nums[NUM_SIDES] = 6;
    dice_nums[MULTIPLIER] = 1;
    dice_nums[MODIFIER] = 0;
    dice_nums[NUM_DROP] = 0;

    while(*dice_string != '\0') {
	if( isdigit(*dice_string) ) {
	    sscanf(dice_string, "%d", &temp_int);
	    while(isdigit(*(++dice_string)));
	}
	else {
	    switch(*dice_string) {
	    case 'd':
		if(temp_int > 0 && temp_int < MAXSHORT)
		    dice_nums[NUM_DICE] = temp_int;
		dice_string++;
		if(*dice_string == '%')
		    dice_nums[NUM_SIDES] = 100;
		else if( (sscanf(dice_string, "%d", &temp_int) < 1 ) ||
			 (temp_int < 2) || (temp_int >= MAXSHORT) ) {
		    free(dice_nums);
		    return NULL;
		} else dice_nums[NUM_SIDES] = temp_int;
		break;
	    case 's':
		if( (sscanf(++dice_string, "%d", &temp_int) < 1) ||
		    (temp_int < 0) || (temp_int >= MAXSHORT)) {
		    free(dice_nums);
		    return NULL;
		} else dice_nums[NUM_DROP] = temp_int;
		break;
	    case 'x':
		if( ( temp_int < 1 ) || (temp_int >= MAXSHORT) ) {
		    free(dice_nums);
		    return NULL;
		} else dice_nums[NUM_ROLLS] = temp_int;
		dice_string++;
		break;
	    case '*':
		if( (sscanf(++dice_string, "%d", &temp_int) < 1) ||
		    (temp_int < 0) || (temp_int >= MAXSHORT)) {
		    free(dice_nums);
		    return NULL;
		} else dice_nums[MULTIPLIER] = temp_int;
		break;
	    case '+':
		if( (sscanf(++dice_string, "%d", &temp_int) < 1) ||
		    (temp_int < 0) || (temp_int >= MAXSHORT)) {
		    free(dice_nums);
		    return NULL;
		} else dice_nums[MODIFIER] = temp_int;
		break;
	    case '-':
		if( (sscanf(++dice_string, "%d", &temp_int) < 1) ||
		    (temp_int < 0) || (temp_int >= MAXSHORT)) {
		    free(dice_nums);
		    return NULL;
		} else dice_nums[MODIFIER] = -temp_int;
		break;
	    default:
		dice_string++;
		break;
	    }
	    temp_int = 0;
	}
    }
    
    return dice_nums;
}




